### PR TITLE
Reduce tcp port usage on master.

### PR DESCRIPTION
**Pull request by xiong-gang**
QD will establish TCP connections to each QE. QEs on the same segment have the same port and
address, thus, we need QE of different gang to have different QD port so that the connection
is a unique <source address, source port, destination address, destination port> quadruple.

To reduce the usage of TCP port on master. We could let the QEs of the same gang on different
segment share on same QD port.

### CODE REVIEW

**Comment by hlinnaka** [comment]
I don't see how forcing a source port helps. The OS is quite capable of reusing the same source port for different target hosts, if you just leave it alone.


On 28 March 2016 11:18:19 EEST, xiong-gang <notifications@github.com> wrote:
>QD will establish TCP connections to each QE. QEs on the same segment
>have the same port and
>address, thus, we need QE of different gang to have different QD port
>so that the connection
>is a unique <source address, source port, destination address,
>destination port> quadruple.
>
>To reduce the usage of TCP port on master. We could let the QEs of the
>same gang on different
>segment share on same QD port.
>You can view, comment on, or merge this pull request online at:
>
>  https://github.com/greenplum-db/gpdb/pull/577
>
>-- Commit Summary --
>
>  * Reduce tcp port usage on master.
>
>-- File Changes --
>
>    M src/backend/cdb/cdbconn.c (18)
>    M src/backend/cdb/cdbgang.c (93)
>    M src/backend/gp_libpq_fe/fe-connect.c (50)
>    M src/backend/gp_libpq_fe/gp-libpq-int.h (2)
>    M src/include/cdb/cdbconn.h (3)
>    M src/include/cdb/cdbgang.h (10)
>
>-- Patch Links --
>
>https://github.com/greenplum-db/gpdb/pull/577.patch
>https://github.com/greenplum-db/gpdb/pull/577.diff
>
>---
>You are receiving this because you are subscribed to this thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/greenplum-db/gpdb/pull/577

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

File: src/backend/gp_libpq_fe/fe-connect.c [hunk 1]
```diff
@@ -1792,6 +1799,40 @@ PQconnectPoll(PGconn *conn)
...
+							continue;
+						}
+
+						qdport = atoi(conn->qdport);
+						Assert(qdport > 0);
+
+						sa_loc.sin_family = addr_cur->ai_family;
+						sa_loc.sin_port = qdport;
+						if (bind(conn->sock, (struct sockaddr *)&sa_loc, sizeof(struct sockaddr)) == -1)
+						{
```
Comment (lij55) [inline src/backend/gp_libpq_fe/fe-connect.c]: Should we handle EAGAIN or EINPROGRESS here?

File: src/backend/cdb/cdbgang.c [hunk 2]
```diff
@@ -796,9 +802,87 @@ buildGangDefinition(GangType type, int gang_id, int size, int content, char *por
...
+	getAvailableTCPPortForGang(&newGangDefinition->dummySock, &newGangDefinition->qdPort);
+
 	return newGangDefinition;
 }
 
+static void 
+getAvailableTCPPortForGang(int *sock, uint16 *port)
+{
+    	int errnoSave = 0;
+	int fd = 0;
```
Comment (lij55) [inline src/backend/cdb/cdbgang.c]: fd shouLd be initialized as -1 because 0 is STDIN

File: src/backend/cdb/cdbgang.c [hunk 3]
```diff
@@ -796,9 +802,87 @@ buildGangDefinition(GangType type, int gang_id, int size, int content, char *por
...
+	int s = 0;
+
+	snprintf(service,32,"%d",0);
+	memset(&hints, 0, sizeof(struct addrinfo));
+	hints.ai_family = AF_UNSPEC;    	/* Allow IPv4 or IPv6 */
+	hints.ai_socktype = SOCK_STREAM; 	/* TCP socket */
+	hints.ai_flags = AI_PASSIVE;    	/* For wildcard IP address */
+	hints.ai_protocol = 0;          	/* Any protocol */
+	
+	if ((s = getaddrinfo(NULL, service, &hints, &addrs)) != 0)
```
Comment (lij55) [inline src/backend/cdb/cdbgang.c]: What will happen here if service is set to "0"?

**Comment by xiong-gang** [comment]
@hlinnaka 
Currently the port is not reused, or else customer won't run of master port. I tried to write some test code, also don't how to let OS handle it.
Besides, there're usually more than one segment on one host, which means, the connections have the same target host.

File: src/backend/cdb/cdbgang.c (hunk 3 above)
Comment (xiong-gang) [inline src/backend/cdb/cdbgang.c]: gotcha

File: src/backend/cdb/cdbgang.c (hunk 2 above)
Comment (xiong-gang) [inline src/backend/cdb/cdbgang.c]: gotcha

File: src/backend/gp_libpq_fe/fe-connect.c (hunk 1 above)
Comment (xiong-gang) [inline src/backend/gp_libpq_fe/fe-connect.c]: i don't think they have such error number

**Comment by hlinnaka** [comment]
I think setting SO_REUSEADDR before connect() is the key here. All the other stuff is unnecessary.

On 28 March 2016 12:39:27 EEST, xiong-gang <notifications@github.com> wrote:
>@hlinnaka 
>Currently the port is not reused, or else customer won't run of master
>port. I tried to write some test code, also don't how to let OS handle
>it.
>Besides, there're usually more than one segment on one host, which
>means, the connections have the same target host.
>
>
>---
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/greenplum-db/gpdb/pull/577#issuecomment-202322715

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by xiong-gang** [comment]
@hlinnaka 
tested and it works. Changed the code.
Thanks.

**Comment by ascherbaum-pivotal** [comment]
@xiong-gang Did you update your profile after submitting this PR, or did the bot not pick up that you are a team member?

File: src/backend/gp_libpq_fe/fe-connect.c [hunk 4]
```diff
@@ -1743,6 +1744,14 @@ PQconnectPoll(PGconn *conn)
 #endif /* SIO_KEEPALIVE_VALS */
 #endif /* WIN32 */
 
+						if (setsockopt(conn->sock, SOL_SOCKET, SO_REUSEADDR, &reuse, sizeof(int)) == -1)
+						{
+							appendPQExpBuffer(&conn->errorMessage, 
+									  libpq_gettext("setsockopt(SO_REUSEADDR) failed:: %s\n"), 
+									  SOCK_STRERROR(SOCK_ERRNO, sebuf, sizeof(sebuf)));
```
Comment (hlinnaka) [inline src/backend/gp_libpq_fe/fe-connect.c]: two colons

**Comment by hlinnaka** [comment]
Ok, LGTM.

**Comment by hlinnaka** [comment]
Oh, one more thing: please add comment explaining why the SO_REUSEADDR is needed.

**Comment by danielgustafsson** [comment]
> On 01 Apr 2016, at 16:19, Andreas Scherbaum <notifications@github.com> wrote:
> 
> @xiong-gang Did you update your profile after submitting this PR, or did the bot not pick up that you are a team member?

The greenplum affiliation needs to be made public, see description on this PR: https://github.com/greenplum-db/gpdb/pull/583#issuecomment-205160238

**Comment by danielgustafsson** [comment]
with @hlinnaka's comments above addressed, :+1:

**Comment by xiong-gang** [comment]
pushed.
Thanks for the review.

**Pull request closed without merging**